Add new option for RightCurlyCheck, issue #1019.

MEZk opened the pull request:
@romani 
I've added new option "alone_or_singleline" for RightCurlyCheck to enforce "{" to be alone on line and allow [single-line-single-statement](https://github.com/MEZk/checkstyle/commit/a756f4db8c07505aaead9cd6df97b59182f0193c#diff-78a4d0ad2378fd234b5628cb24fcd469R41) format of block. I've also provided UTs and generated [reports](http://mezk.github.io/index.html) to see the difference between ALONE and ALONE_OR_SINGLELINE options. As you can see from the reports, new option makes sense, single-line-sigle-statement blocks are often used in methods and constructors (especially in hibernate and apache-ant).

romani commented:
while I am reviewing code ....

1) Please enforce 100% coverage on methods/code that you created.

2) Guava reports, please redo them to match configuration that is Google_style.xml
https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L63
https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L64

Please for all project generate report before update  for ALONE value, it should match to ALONE after update. We need to be sure that existing mode work exactly was it was before (or better :) ).

MEZk commented:
@romani 
1)
>Please enforce 100% coverage on methods/code that you created.
 
Done.
2) 
>Guava reports, please redo them to match configuration that is Google_style.xml
https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L63
https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L64
Please for all project generate report before update for ALONE value, it should match to ALONE after update. We need to be sure that existing mode work exactly was it was before (or better :) ).

Done. See [REPORTS] (http://mezk.github.io/index.html)

3) I've also found [dead code] (https://github.com/checkstyle/checkstyle/commit/087a6b7bd6d7fba303ee5f0ff779315f8199b94c#diff-9b2a5f9e69b254b923cb4fd7420cb9a4R334). We do not need to check
 ````java
if (rightCurly != null)
````
in 'hasLineBreakBefore' because when we invoke 'hasLineBreakBefore'  it always  is not null due to the fact that we check
````java
if (rcurly == null || rcurly.getType() != TokenTypes.RCURLY)
````
in [visitToken](https://github.com/checkstyle/checkstyle/commit/087a6b7bd6d7fba303ee5f0ff779315f8199b94c#diff-9b2a5f9e69b254b923cb4fd7420cb9a4R152) and then we invoke 'hasLineBreakBefore' only in [validate](https://github.com/checkstyle/checkstyle/commit/087a6b7bd6d7fba303ee5f0ff779315f8199b94c#diff-9b2a5f9e69b254b923cb4fd7420cb9a4R184).

4) I also think that 
````java
while (parent != null ...
````
in [getNextToken](https://github.com/checkstyle/checkstyle/commit/087a6b7bd6d7fba303ee5f0ff779315f8199b94c#diff-9b2a5f9e69b254b923cb4fd7420cb9a4R321) is a dead code.
I'll prove these predictions but in another PR.

romani commented:
> I've also found dead code. We do not need to check

Main point is do not introduce dead code. If code existed before your update, it could be fixed in next PRs.

romani commented:
1) please split into two commits of the same PR : formatting and typos fixes , issue fix. That will ease code review.

2) please add example of usage at http://checkstyle.sourceforge.net/config_blocks.html#RightCurly 

3) 
> isSingleStatementBlock

I think we should just Check for single line by new option (not a single like and single statement), as it Checking for single statement per line is another Check and it could be switched OFF at user config (http://checkstyle.sourceforge.net/config_coding.html#OneStatementPerLine). Mixing few policies in one Check is not good, as user lose ability to adjust Checkstyle to his needs.

MEZk commented:
@romani 
3)
>I think we should just Check for single line by new option (not a single like and single statement), as it Checking for single statement per line is another Check and it could be switched OFF at user config (http://checkstyle.sourceforge.net/config_coding.html#OneStatementPerLine). Mixing few policies in one Check is not good, as user lose ability to adjust Checkstyle to his needs.

If we check  for single line by new option it means that this will be allowed:
````java
private void foo1() {int var1 = 5; var2 = 6; }

private void foo2()
 {int var1 = 5; var2 = 6; }
````
because we will check that the block ```{ //some logic }``` is on a single line. 

But this will not be allowed:

````java
private void foo1() { int var1 = 5; var2 = 6; } private void foo2() {int var1 = 5; var2 = 6; }
````

because after first block ```{int var1 = 5; var2 = 6; } ``` there is the next token ```METHOD_DEF```
Is it right?

In this case user can use OneStatementPerLine to check that inside block there is only one statement per line.

romani commented:
> If we check for single line by new option it means that this will be allowed

yes, it is expected.

> But this will not be allowed:

yes, it is expected

> Does it make sense ?

yes it make, code looks weird only because to make it good you need to use LeftCurly and OneStatementPerLine together with RightCurly. But each of Check is independent and could be used even alone if user wish to have that.

MEZk commented:
1) I've reimplemented new logic and deleted unnecessary code according to above conversation. I tried to imagine all possible and impossible :) examples in tests.You can see additional test inputs [here](https://github.com/MEZk/checkstyle/commit/608760c33d06cddf04b21b681ed07a68516f90aa#diff-78a4d0ad2378fd234b5628cb24fcd469R9). I've already covered all new logic except my predictions about dead code above. 

2) I returned typos, I'll correct them in another PR;

3)
>
As you reimplement new logic, try to compare reports of previous implementation and new, to see is it really become better .

It is really become better, because now this code rises violation with new option:

````java
}});
````
And there are many similar examples in guava's code.

4) All reports are [here](http://mezk.github.io).

[6.8.1-RELEASE (option = alone)](http://mezk.github.io/guava_6.8.1_alone/checkstyle.html) (102 violation)
 and 
[6.9-SNAPSHOT (option = alone)](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/checkstyle.html) (102 violation) are equal, proved with MELD :+1:

romani commented:
very clear analysis , thanks a lot.

1) 
http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-testlib/src/com/google/common/testing/GcFinalization.html#L194
http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L580

false positive, all "}" are on the same line as "{". Please be attentive, you do not have hundreds of violations, you can review them all.

2)
http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/collect/MapsTest.html#L189

Just FYI : that is very interesting cases .... I do not know how to make it formatted better. But I do not think we should support that case at least for now.

3)
```java
                && (rcurly.getLineNo() == nextToken.getLineNo()
                && nextToken.getType() != TokenTypes.RCURLY
                || rcurly.getLineNo() == lcurly.getLineNo())
                && !isEmptyBody(lcurly)) {
```

second and third rows should be intended further to the right as they are not in the same level of execution as fist and last "&&"

MEZk commented:
@romani 
1)
>http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-testlib/src/com/google/common/testing/GcFinalization.html#L194

I though that there should be only one single-line block per line including inner single-line blocks. I fixed this misunderstanding. And added addition [test input](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-78a4d0ad2378fd234b5628cb24fcd469R116) for [testAloneOrSingleLine] (https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-28445257cef7838e0afb59143ffe0738R194).

>http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L580

Hm... It is not a problem of the new option. I checked all reports. It is due to the fact that I use the following configuration with checkstyle-tester:

````xml
        <module name="RightCurly"/>
        <module name="RightCurly">
            <property name="option" value="alone_or_singleline"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
        </module>
````

By default right_curly_policy for RightCurly module is RightCurlyOption.SAME. This means that the right curly must be on the same line as the next statement. 

Lets analyse the [following code example](http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L580):

````java
 } finally { future.cancel(true); }
````

To begin with, first ```}``` token. right_curly_policy is RightCurlyOption.SAME. It is ok that after ```}``` is ```finally``` statement. All right. 

Next, right_curly_policy is RightCurlyOption.SAME, but look at the second ```}```. There is no statement after second ```}```. That is why with  RightCurlyOption.SAME the Check rises violation here that ```	'}' should have line break before.``` And the same violation is at [6.8.1-RELEASE (102 violations)](http://mezk.github.io/guava_6.8.1_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L580) and [6.9-SNAPSHOT (102 violations)](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L580). If you think that this is a bug, I'll fix it as a seperate issue.

What is more, as you can see from [Rules](http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/checkstyle.html), token LITERAL_FINALLY is not specified for RightCurly module. This means that there can not be a violation caused by new option. Nevertheless, I add this test input at [InputRightCurlyWithAnnotation](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-78a4d0ad2378fd234b5628cb24fcd469R126) and specified all [acceptableTokens](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-28445257cef7838e0afb59143ffe0738R196) for current check at [testAloneOrSingleLine](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-28445257cef7838e0afb59143ffe0738R194) and [testWithAnnotations](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-28445257cef7838e0afb59143ffe0738R153). 

2)
>http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/collect/MapsTest.html#L189
Just FYI : that is very interesting cases .... I do not know how to make it formatted better. But I do not think we should support that case at least for now.

Fixed.

I generated [report](http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_fix/checkstyle.html) with the following configuration after all fixes:

````xml
        <module name="RightCurly">
            <property name="option" value="alone_or_singleline"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
        </module>
````

You may think that there is a very small amount (1) of violation but it is due to the fact that not all tokens were specified in module configuration.

3) Corrected.

MEZk commented:
@romani 
rebased

romani commented:
appveyor is passing for master, please recommit smth in branch to re-trigger CIs builds. From this moment I expect all CIs to be green.

romani commented:
1)
> And the same violation is at 6.8.1-RELEASE (102 violations) and 6.9-SNAPSHOT (102 violations). If you think that this is a bug, I'll fix it as a separate issue.

as such behavior was before, please create separate issue.

2)
> And there should be violation with the new option, because there is LITERAL_FINALLY token after first } and that is why first } is not alone on the line, and also all block before finally is not on a single-line.

User will never succeeded to make such an analysis, please extend message with reporting a position of wrong "}" . That will help user a lot, it will help us first of all :) . 

3)
> You may think that there is a very small amount (1) of violation but it is due to the fact that not all tokens were specified in module configuration (only from google_style).

I do not understand you, please try to explain one more time how you fixed a problem.

4)
Please move all formatting to separate PR and send it separately I will apply it immediately. Now we have very complicated issue fix and I do not want to see only functional changes clearly.

MEZk commented:
@romani 
1) Done.
2) It is expected behaviour for alone_or_singleline option, I provided test for that:
[test input](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-78a4d0ad2378fd234b5628cb24fcd469R126)
[test](https://github.com/checkstyle/checkstyle/commit/f6c57be19a552cf4146955ca758388c9337a79f8#diff-28445257cef7838e0afb59143ffe0738R207)
Line and column number is included in output message:
```"126:9: "}" should be alone on a line.```
3) I allowed [this](http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-testlib/src/com/google/common/testing/GcFinalization.html#L194) and [this](http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/collect/MapsTest.html#L189) cases and it decreased the amount of violations. And I did not use the full set of acceptable tokens for current check when I generated reports with checkstyle-tester that is why LITERAL_FINALLY, LITERAL_TRY, LITERAL_CATCH and other literals were not checked.
4) Done.

romani commented:
>  I allowed this and this cases and it decreased the amount of violations.
http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline_2/xref/guava/guava-tests/test/com/google/common/collect/MapsTest.html#L189

I do not see that test case in Inputs of UTs. Each problematic case should be in UTs, I do not want to do detailed guava report analysis on each change in that Check. If I ma blind please send me a link

MEZk commented:
@romani 
Sorry, added [this](https://github.com/checkstyle/checkstyle/commit/2ee7fa9d9746c1af439c2d850322497971017f63#diff-78a4d0ad2378fd234b5628cb24fcd469R129) and [this](https://github.com/checkstyle/checkstyle/commit/2ee7fa9d9746c1af439c2d850322497971017f63#diff-78a4d0ad2378fd234b5628cb24fcd469R116) as  test inputs

romani commented:
> Please send me a link to condition that allow such cases to pass validation.

please do

MEZk commented:
@romani 
[here](https://github.com/MEZk/checkstyle/commit/2ee7fa9d9746c1af439c2d850322497971017f63#diff-9b2a5f9e69b254b923cb4fd7420cb9a4R205)

romani commented:
By means of that condition you ignore all sequential right curly braces , but I asked you to allow only at intance initalization block.
After fix please send me links to test case where double right curly are not ok.

romani commented:
@MEZk , let finish that issue , till you and me remember context of it.

MEZk commented:
@romani 

Links to test cases where double right curly are not ok:
[1](https://github.com/checkstyle/checkstyle/commit/864d6fd83f6d174abbc4202354b4a9d685f9423d#diff-78a4d0ad2378fd234b5628cb24fcd469R192)
[2](https://github.com/checkstyle/checkstyle/commit/864d6fd83f6d174abbc4202354b4a9d685f9423d#diff-78a4d0ad2378fd234b5628cb24fcd469R189)
[3](https://github.com/checkstyle/checkstyle/commit/864d6fd83f6d174abbc4202354b4a9d685f9423d#diff-78a4d0ad2378fd234b5628cb24fcd469R185)
[4](https://github.com/checkstyle/checkstyle/commit/864d6fd83f6d174abbc4202354b4a9d685f9423d#diff-78a4d0ad2378fd234b5628cb24fcd469R175)
[5](https://github.com/checkstyle/checkstyle/commit/864d6fd83f6d174abbc4202354b4a9d685f9423d#diff-78a4d0ad2378fd234b5628cb24fcd469R153)

Rises violation only for ALONE:
[6](https://github.com/checkstyle/checkstyle/commit/864d6fd83f6d174abbc4202354b4a9d685f9423d#diff-78a4d0ad2378fd234b5628cb24fcd469R116)

Fixed mistakes in tests:
1) testDefault and testSame

Deleted: 
````java
"93:27: " + getCheckMessage(MSG_KEY_LINE_NEW, "}"),
"93:27: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}"),
````
Reason: testDefault and testSame use right curly policy SAME.  That is why ```}``` on [line 93](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputLeftCurlyOther.java#L93) should be on the same line as the next statement.

2) testAlone

Deleted: 
````java
"93:27: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}"),
````
Reason: testAlone uses right curly policy ALONE.  That is why ```}``` on [line 93](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputLeftCurlyOther.java#L93) should be on alone on a line.

3) testNewLine

Added:
````java
"111:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}"),
 "122:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}"),
````
Reason: testNewLine uses right curly policy ALONE.  That is why the first```}``` on [line 111](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputLeftCurlyOther.java#L111) should be on alone on a line and the second ```}``` should be on a new line. The first```}``` on [line 122](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputLeftCurlyOther.java#L122) should be on alone on a line and the second ```}``` should be on a new line. 

All mistakes were due to the fact that there were two violations: [first](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L178) was inside validate method and [second](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L168) was inside visitToken itself. So, if some code rose violation inside validate method while checking then it might also rise violation in visitToken method.

After all fixes guava_6.8.1_alone (102 violation) and guava_6.9-SNAPSHOT_alone (121 violation) have 19 differences:

| №        | Line          |Class | Reason  |
| ------------- |:-------------:| -----:| -----:|
| 1   | [128](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-gwt/test-super/com/google/common/collect/super/com/google/common/collect/MapsTest.html#L128) | MapsTest | ```}}``` on the same line|
| 2   | [91](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-testlib/src/com/google/common/util/concurrent/testing/AbstractListenableFutureTest.html#L91) | AbstractListenableFutureTest | ```}}``` on the same line|
| 3   | [108](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-testlib/test/com/google/common/testing/GcFinalizationTest.html#L108) | GcFinalizationTest | ```}}}``` on the same line|
| 4   | [178](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/benchmark/com/google/common/util/concurrent/ExecutionListBenchmark.html#L178) | ExecutionListBenchmark |```}}``` on the same line|
| 5   | [504](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/benchmark/com/google/common/util/concurrent/ExecutionListBenchmark.html#L504) | ExecutionListBenchmark |```}}``` on the same line|
| 6   | [32](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/cache/PackageSanityTests.html#L32) | PackageSaintyTest |```}}``` on the same line|
| 7   | [189](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/collect/MapsTest.html#L189) | MapsTest |```}}``` on the same line|
| 8   | [429](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/AbstractScheduledServiceTest.html#L429) | AbstractDoubleArrayTest |```}}``` on the same line|
| 9   | [474](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/AbstractScheduledServiceTest.html#L474) | AbstractDoubleArrayTest |```}}``` on the same line|
| 10   | [209](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/AtomicDoubleArrayTest.html#L209) | AtomicDoubleArrayTest |```}}``` on the same line|
| 11   | [125](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/AtomicDoubleTest.html#L125) | AtomicDoubleTest |```}}``` on the same line|
| 12   | [919](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L919) | JSR166TestCase |```}}``` on the same line|
| 13   | [926](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L926) | JSR166TestCase |```}}``` on the same line|
| 14   | [1023](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.html#L1023) | JSR166TestCase |```}}``` on the same line|
| 15   | [202](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/MoreExecutorsTest.html#L202) | MoreExecutorsTest |```}}``` on the same line|
| 16   | [600](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/MoreExecutorsTest.html#L600) | MoreExecutorsTest |```}}``` on the same line|
| 17   | [317](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava-tests/test/com/google/common/util/concurrent/ServiceManagerTest.html#L317) | ServiceManagerTest |```}}``` on the same line|
| 18   | [332](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava/src/com/google/common/cache/Striped64.html#L332) | Striped64 |```}}``` on the same line|
| 19   | [362](http://mezk.github.io/guava_6.9-SNAPSHOT_alone/xref/guava/guava/src/com/google/common/primitives/UnsignedBytes.html#L362) | UnsignedBytes |```}}``` on the same line|

[REPORTS](http://mezk.github.io)
[Double Brace Initialization info](http://c2.com/cgi/wiki?DoubleBraceInitialization)

romani commented:
1) 
> flag = !flag; } System.err.
              println("Xe-xe");

Such code should not be OK for any option! Non of IDEs will support that crap.
Root of the problem is at http://checkstyle.sourceforge.net/property_types.html#rcurly " 	The brace must be on the same line as the next statement." here 'statement' mean next part of multi-block statements (http://google.github.io/styleguide/javaguide.html#s4.1.3-braces-empty-blocks)

2)

> but I asked you to allow only at instance initialization block.

Explanation of Alone option:
http://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.1.2-blocks-k-r-style
So we have to be strict with "alone" mode so instance initialization should be marked as violation, as you have it in your report already.
If users start complaining about that - we will introduce new option smth like "waiverParentToken" to let users define that instance initialization braces should be skipped. But that should be done as different issue.

3)
http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline/checkstyle.html

> guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.java
Severity	Category	Rule	Message	Line
 Warning	blocks	RightCurly	'}' should have line break before.	580

It is not clear to what curly brace is violated. We need to print user a column number to exactly point him to a problem.

4)
http://mezk.github.io/guava_6.9-SNAPSHOT_alone_or_singleline/checkstyle.html

>guava/guava/src/com/google/common/cache/Striped64.java
Severity	Category	Rule	Message	Line
 Warning	blocks	RightCurly	'}' should be on the same line.	199

I do not understand reason of that violation , brace is alone  ,what is a problem with it ?

5)
In next report please provide explanation to each violation.

romani commented:
@MEZk , can you focus on that PRs

MEZk commented:
@romani 
1)
>Such code should not be OK for any option! Non of IDEs will support that crap.
Root of the problem is at http://checkstyle.sourceforge.net/property_types.html#rcurly " The brace must be on the same line as the next statement." here 'statement' mean next part of multi-block statements (http://google.github.io/styleguide/javaguide.html#s4.1.3-braces-empty-blocks)

That code rises violation.
SAME: [UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R65), [test input](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputLeftCurlyOther.java#L93)
ALONE: [UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R75), [test input](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputLeftCurlyOther.java#L93)
ALONE_OR_SINGLELINE: [UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R236), [test input](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-78a4d0ad2378fd234b5628cb24fcd469R207)

2)
>Explanation of Alone option:
http://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.1.2-blocks-k-r-style
So we have to be strict with "alone" mode so instance initialization should be marked as violation, as you have it in your report already.
If users start complaining about that - we will introduce new option smth like "waiverParentToken" to let users define that instance initialization braces should be skipped. But that should be done as different issue.

Instance initialization rises violation when ALONE option is used.
[UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R182), [test input1](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-78a4d0ad2378fd234b5628cb24fcd469R116)
[UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R186), [test input2](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-78a4d0ad2378fd234b5628cb24fcd469R137)
[UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R187), [test input3](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-78a4d0ad2378fd234b5628cb24fcd469R138)
[UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R189), [test input4](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-78a4d0ad2378fd234b5628cb24fcd469R139)
[UT](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-28445257cef7838e0afb59143ffe0738R190), [test input5](https://github.com/checkstyle/checkstyle/commit/18b7c3978a17687c4ce78836579e61652400ba23#diff-78a4d0ad2378fd234b5628cb24fcd469R140)

3)
>It is not clear to what curly brace is violated. We need to print user a column number to exactly point him to a problem.

In this case we should also do refactoring for LeftCurlyCheck and it's UTs because LefctCurlyCheck uses the same messages from property file or create new properties for RightCurlyCheck. Will it be ok?

4)
>I do not understand reason of that violation , brace is alone ,what is a problem with it ?

It is a bug of RightCurlyOption.SAME, issue #1416. When ```}``` ends a line, the Check rises violation because ```}``` should be on the same line as the next statement.

romani commented:
https://travis-ci.org/checkstyle/checkstyle/jobs/73143789#L278

remove unused imports, in Idea cntl+alt+o.

MEZk commented:
@romani 
Correted

romani commented:
> Will it be ok, because the current PR is already big?

agree, please another issue.

MEZk commented:
@romani 
New issue - [#1511](https://github.com/checkstyle/checkstyle/issues/1511)
Please, see 1) and 2) from my previous comment.
I've separated files with test inputs into two files: one for test with annotations and one for test with new option and marked violations.

romani commented:
merged as FF, thanks a lot for huge update
